pkg/eksd/upgrader.go

```diff
@@ -23,34 +25,40 @@ func NewUpgrader(client EksdInstallerClient, reader Reader, opts ...UpgraderOpt)
 	}
 }
 
-func (u *Upgrader) Upgrade(ctx context.Context, cluster *types.Cluster, currentSpec, newSpec *cluster.Spec) (*types.ChangeDiff, error) {
+// Upgrade checks whether upgrading EKS-D components is necessary and, if so, installs the new EKS-D CRDs.
+func (u *Upgrader) Upgrade(ctx context.Context, cluster *types.Cluster, currentManagementComponentsVersionsBundle *releasev1alpha1.VersionsBundle, newSpec *cluster.Spec) (*types.ChangeDiff, error) {
 	logger.V(1).Info("Checking for EKS-D components upgrade")
-	changeDiff := EksdChangeDiff(currentSpec, newSpec)
+
+	changeDiff := ChangeDiff(currentManagementComponentsVersionsBundle, &newSpec.Bundles.Spec.VersionsBundles[0])
 	if changeDiff == nil {
 		logger.V(1).Info("Nothing to upgrade for EKS-D components")
 		return nil, nil
 	}
 	logger.V(1).Info("Starting EKS-D components upgrade")
 	if err := u.InstallEksdCRDs(ctx, newSpec, cluster); err != nil {
-		return nil, fmt.Errorf("upgrading EKS-D components from version %s to version %s: %v", changeDiff.ComponentReports[0].OldVersion, changeDiff.ComponentReports[0].NewVersion, err)
+		return nil, fmt.Errorf("upgrading EKS-D components to EKS-A version %s: %v", *newSpec.Cluster.Spec.EksaVersion, err)
```

Comment:
I'm not against this but do we have any validation that protects against nil eksa version?
if not, dereferencing it like this it's dangerous